pgmax/fg/groups.py

```diff
@@ -151,255 +96,60 @@ def unflatten(self, flat_data: Union[np.ndarray, jnp.ndarray]) -> Any:
         factor_configs: Optional array containing an explicit enumeration of all valid configurations
         log_potentials: Array of log potentials.
 
     Attributes:
         factor_type: Factor type shared by all the Factors in the FactorGroup.
         factor_sizes: Array of the different factor sizes.
-        variables_for_factors: Tuple concatenating the variables connected to each factor in the FactorGroup.
-            Each variable will appear once for each Factor it connects to.
         factor_edges_num_states: Array concatenating the number of states for the variables connected to each Factor of
             the FactorGroup. Each variable will appear once for each Factor it connects to.
 
     Raises:
         ValueError: if the FactorGroup does not contain a Factor
     """
 
-    variable_group: Union[CompositeVariableGroup, VariableGroup]
-    variable_names_for_factors: Sequence[List]
+    variables_for_factors: Sequence[List]
     factor_configs: np.ndarray = field(init=False)
     log_potentials: np.ndarray = field(init=False, default=np.empty((0,)))
     factor_type: Type = field(init=False)
     factor_sizes: np.ndarray = field(init=False)
-    variables_for_factors: Tuple[nodes.Variable, ...] = field(init=False)
     factor_edges_num_states: np.ndarray = field(init=False)
 
     def __post_init__(self):
-        if len(self.variable_names_for_factors) == 0:
-            raise ValueError("Do not add a factor group with no factors.")
-
-        factor_sizes = []
-        variables_for_factors = []
-        factor_edges_num_states = []
-        for variable_names_for_factor in self.variable_names_for_factors:
-            for variable_name in variable_names_for_factor:
-                variable = self.variable_group._names_to_variables[variable_name]
-                variables_for_factors.append(variable)
-                factor_edges_num_states.append(variable.num_states)
-            factor_sizes.append(len(variable_names_for_factor))
-
-        object.__setattr__(self, "factor_sizes", np.array(factor_sizes))
-        object.__setattr__(self, "variables_for_factors", tuple(variables_for_factors))
-        object.__setattr__(
-            self, "factor_edges_num_states", np.array(factor_edges_num_states)
+        if len(self.variables_for_factors) == 0:
+            raise ValueError("Cannot create a FactorGroup with no Factor.")
+
+        factor_sizes = np.array(
+            [
+                len(variables_for_factor)
+                for variables_for_factor in self.variables_for_factors
+            ]
+        )
+        object.__setattr__(self, "factor_sizes", factor_sizes)
+
+        factor_edges_num_states = np.empty(shape=(self.factor_sizes.sum(),), dtype=int)
+        _compile_edges_num_states_numba(
+            factor_edges_num_states, self.variables_for_factors
         )
+        object.__setattr__(self, "factor_edges_num_states", factor_edges_num_states)
```

Comment:
These are pre-computed to speed things up? Can we make then cached properties instead of attributes?